SDK 1.4.1 integration

kpyszkowski opened the pull request:
Ref: https://github.com/threshold-network/token-dashboard/issues/647

- Integrated SDK 1.4.1 with dApp
- Updated project to support Node 18
- Tested build process for `goerli` and `development` packages
- Bumped Node version in CI/CD process configuration files

Note:
- There is a brand new warning  while `yarn install`ing ✨
  It ain't a blocker and I couldn't pass through it easily so I decided to leave it for now.

lukasz-zimnoch commented on package.json:
```diff
@@ -2,6 +2,9 @@
   "version": "1.12.0-pre",
   "private": true,
+  "engines": {
+    "node": "<=16"
```
I think this should be:
```suggestion
    "node": ">=16"
```

kpyszkowski replied:
Ref commit: 76ea1249e16cfb691617ad6a927f3ab46776b47e

lukasz-zimnoch commented on package.json:
```diff
@@ -55,7 +58,7 @@
     "react-confetti": "^6.0.1",
     "react-dom": "^17.0.2",
-    "react-icons": "^4.3.1",
+    "react-icons": "<=4.3.1",
```
What's the reason behind this change?

kpyszkowski replied:
It turned out`react-icons` library maintainers didn't follow semantic versioning guidelines shipping major breaking changes with minor version. Prefixing version tag with `^` instructs package manager to download latest minor version which currently has breaking changes - reorganized imports which refuses Chakra UI types generation to complete resulting with error. I concidered refactoring our code to include library updates but it's used in dependencies (`@threshold-network/components`) and it might collide on the run/build-time - needs more analysis. Maybe with post-integration chores?

The latest version without breaking changes is [`4.8.0`](https://github.com/react-icons/react-icons/commit/686e260460a3a4360355548c2f1fc7ca923e95aa). I updated library version resolution limiting up to `4.8.0`'s patches. It's more precise, consistent with current codebase, possibly up-to-date and doesn't collide with dependencies.

lukasz-zimnoch replied:
Ok makes sense! Thanks for the clarification!

lukasz-zimnoch commented on .nvmrc:
```diff
@@ -0,0 +1 @@
+18
```
👌

lukasz-zimnoch commented on package.json:
```diff
@@ -78,7 +81,7 @@
     "chromatic": "npx chromatic --exit-zero-on-changes",
     "cypress:open": "cypress open",
-    "gen:theme-typings": "chakra-cli tokens ./src/theme/index.ts",
+    "gen:theme-typings": "chakra-cli tokens ./src/theme/index.ts --out node_modules/@chakra-ui/styled-system/dist/declarations/src/theming.types.d.ts",
```
Can you say more about this?

kpyszkowski replied:
With Node version bump it was mandatory to update `@chakra-ui/cli` library. Currently it's used [unreleased version with some unknown fixes](https://github.com/chakra-ui/chakra-ui/issues/5219#issuecomment-1001203749) provided by a contributor and it crashed with typescript/node error.
For some unknown reasons the latest version of the CLI doesn't detect `@chakra-ui/styled-system` as a dependency and refuses for theme types generation:
```bash
❯ yarn gen:theme-typings
yarn run v1.22.19
$ chakra-cli tokens ./src/theme/index.ts

 Chakra UI CLI  v2.4.1 by Chakra UI
Generate theme typings for autocomplete


✖ An error occurred
Could not find @chakra-ui/styled-system in node_modules. Please provide `--out` parameter.
✨  Done in 0.87s.
```
The solution was to define the output location explicitly. My assumption is it scans `package.json`  for `@chakra-ui/styled-system` and returns an error when it's not found.

lukasz-zimnoch replied:
Thanks for clarifying this!

lukasz-zimnoch commented on package.json:
```diff
@@ -65,9 +68,9 @@
   },
   "scripts": {
-    "start": "craco start",
+    "start": "craco --openssl-legacy-provider start",
```
Can you elaborate more about this change?

lukasz-zimnoch commented on src/components/Modal/TbtcMintingConfirmationModal/index.tsx:
```diff
@@ -57,13 +60,19 @@ const TbtcMintingConfirmationModal: FC<TbtcMintingConfirmationModalProps> = ({
     onSuccessfulDepositReveal
   )
-
+  const bitcoinNetwork =
+    supportedChainId === ChainID.Ethereum.toString()
+      ? BitcoinNetwork.Mainnet
+      : BitcoinNetwork.Testnet
```
Can we use https://github.com/threshold-network/token-dashboard/blob/39a0be63ac70994e97118dec87a20884a4ed6b6e/src/utils/getThresholdLib.ts#L13 directly instead of repeating this logic here?

michalsmiarowski replied:
> Can we use
> 
> https://github.com/threshold-network/token-dashboard/blob/39a0be63ac70994e97118dec87a20884a4ed6b6e/src/utils/getThresholdLib.ts#L13
> 
> directly instead of repeating this logic here?

We acutally store that bitcoin config inside our threshold lib which we call from hook few lines above:
https://github.com/threshold-network/token-dashboard/blob/b7f806adcaab9b5bf215c922f5126a625f35061b/src/components/Modal/TbtcMintingConfirmationModal/index.tsx#L53

The `config` is a public property there so I think we can just use:
```
threshold.config.bitcoin.network
```

cc @kpyszkowski

kpyszkowski replied:
Ref commit: 2cb19bd306d30c3d82809bc974b6363b730a65dd

lukasz-zimnoch commented on src/tbtc/mock-bitcoin-client.ts:
```diff
@@ -38,6 +40,11 @@ export const testnetUTXO: UnspentTransactionOutput & RawTransaction = {
   ...testnetTransaction,
 }
 const testnetPrivateKey = "cRJvyxtoggjAm9A94cB86hZ7Y62z2ei5VNJHLksFi2xdnz1GJ6xt"
+const bitcoinNetwork =
+  supportedChainId === ChainID.Ethereum.toString()
+    ? BitcoinNetwork.Mainnet
+    : BitcoinNetwork.Testnet
```
Same regarding https://github.com/threshold-network/token-dashboard/blob/39a0be63ac70994e97118dec87a20884a4ed6b6e/src/utils/getThresholdLib.ts#L13

michalsmiarowski replied:
The `MockBitcoinClient` class is used only to mock transactions/utxos based on testnet values, so I think we should use testnet network here always.

@kpyszkowski You can use `this.getNetwork()` method - as of now it always returns testnet network:
https://github.com/threshold-network/token-dashboard/blob/b7f806adcaab9b5bf215c922f5126a625f35061b/src/tbtc/mock-bitcoin-client.ts#L290-L292

kpyszkowski replied:
Ref commit: 2cb19bd306d30c3d82809bc974b6363b730a65dd

lukasz-zimnoch commented on src/tbtc/mock-bitcoin-client.ts:
```diff
@@ -38,6 +40,11 @@ export const testnetUTXO: UnspentTransactionOutput & RawTransaction = {
+  supportedChainId === ChainID.Ethereum.toString()
+    ? BitcoinNetwork.Mainnet
+    : BitcoinNetwork.Testnet
+const bitcoinNetworkTransactionFee = BigNumber.from("47000")
```
Is this just an arbitrary value? Worth clarifying it with a comment

kpyszkowski replied:
Ref commit: 2cb19bd306d30c3d82809bc974b6363b730a65dd

lukasz-zimnoch commented on src/threshold-ts/tbtc/index.ts:
```diff
@@ -403,6 +405,11 @@ export interface ITBTC {
   }>
 }
 
+const bitcoinNetwork =
+  supportedChainId === ChainID.Ethereum.toString()
+    ? BitcoinNetwork.Mainnet
+    : BitcoinNetwork.Testnet
```
Same regarding https://github.com/threshold-network/token-dashboard/blob/39a0be63ac70994e97118dec87a20884a4ed6b6e/src/utils/getThresholdLib.ts#L13

michalsmiarowski replied:
@kpyszkowski See https://github.com/threshold-network/token-dashboard/pull/652#discussion_r1372887502

We should use `this.bitcoinNetwork` here:
https://github.com/threshold-network/token-dashboard/blob/b7f806adcaab9b5bf215c922f5126a625f35061b/src/threshold-ts/tbtc/index.ts#L483-L485

Example:
https://github.com/threshold-network/token-dashboard/blob/b7f806adcaab9b5bf215c922f5126a625f35061b/src/threshold-ts/tbtc/index.ts#L551-L559

kpyszkowski replied:
Ref commit: 2cb19bd306d30c3d82809bc974b6363b730a65dd

lukasz-zimnoch commented on src/utils/tBTC.ts:
```diff
@@ -5,13 +5,20 @@ import {
 const UNMINTING_MAINNET_BTC_ADDRESS_PREFIXES = ["1", "bc1", "3"] as const
 const UNMINTING_TESTNET_BTC_ADDRESS_PREFIXES = ["m", "n", "tb1", "2"] as const
 
+const bitcoinNetwork =
+  supportedChainId === ChainID.Ethereum.toString()
+    ? BitcoinNetwork.Mainnet
+    : BitcoinNetwork.Testnet
```
Same regarding https://github.com/threshold-network/token-dashboard/blob/39a0be63ac70994e97118dec87a20884a4ed6b6e/src/utils/getThresholdLib.ts#L13

michalsmiarowski replied:
Here, based on my previous comment (https://github.com/threshold-network/token-dashboard/pull/652#discussion_r1372887502) I propose to add additional argument (`btcNetwork`) to `withBitcoinAddress` and to `buildRedemptionDetailsLink` methods and then just get the network from threshold lib and pass it here:
https://github.com/threshold-network/token-dashboard/blob/b7f806adcaab9b5bf215c922f5126a625f35061b/src/components/Modal/tBTC/InitiateUnminting.tsx#L57-L67

kpyszkowski replied:
Ref commit: 2cb19bd306d30c3d82809bc974b6363b730a65dd

kpyszkowski commented on .github/workflows/reusable-build-and-publish.yml:
```diff
@@ -156,7 +156,7 @@ jobs:
             @keep-network/random-beacon@${{ steps.set-packages-versions.outputs.random-beacon-contracts-version }} \
             @keep-network/tbtc-v2@${{ steps.set-packages-versions.outputs.tbtc-v2-contracts-version }} \
-            @keep-network/tbtc-v2.ts@${{ inputs.environment }}
+            @keep-network/tbtc-v2.ts@1.4.1-goerli.0
```
@lukasz-zimnoch 
Regarding to our Discord conversation, please let me know is it what you wanted me to do.

lukasz-zimnoch replied:
Exactly! Let's add a TODO that this line should be removed once the dashboard is migrated to SDK v2

kpyszkowski replied:
Ouch, sorry. Forgot about the comment 🙇🏼 
Ref commit: 6b811a686b9078f324644ef411df576ef1464f04

lukasz-zimnoch commented on src/utils/tBTC.ts:
```diff
@@ -5,13 +5,16 @@ import {
   createOutputScriptFromAddress,
   prependScriptPubKeyByLength,
 } from "../threshold-ts/utils"
+import { getThresholdLib } from "./getThresholdLib"
```
Nitpick: Can't we get the already initialized `threshold` object directly?

lukasz-zimnoch commented on src/threshold-ts/tbtc/index.ts:
```diff
@@ -54,6 +54,7 @@ import {
   findWalletForRedemption,
 } from "@keep-network/tbtc-v2.ts/dist/src/redemption"
 import { TBTCToken as ChainTBTCToken } from "@keep-network/tbtc-v2.ts/dist/src/chain"
+import { getThresholdLib } from "../../utils/getThresholdLib"
```
Unused import?

kpyszkowski replied:
Removed 🙇🏼 
Ref commit: 7baf050573125f36b5e4cea69e025c781de60648

lukasz-zimnoch reviewed (approved):
Last comments! Otherwise, LGTM! Leaving the final approval for @michalsmiarowski. Let's cut a release once this PR is merged and test those changes on mainnet preview.

kpyszkowski commented:
@lukasz-zimnoch Your approval has been canceled, I had to push minor fixes 🙌🏼

michalsmiarowski reviewed (changes requested):
Left one non-blocking comment down below

michalsmiarowski commented:
@kpyszkowski 

It looks like all (or almost all) libraries in `yarn.lock` got updates which IMO is not a good idea as it is easy to miss some bugs and not all libraries could follow semantic versioning, which means not all changes could be non-braking. For example, since we updates prettier, we now had to do the changes in 545a5c32a50eb5a4fbdd123dc9055e35e06d035b, but it looks like we don't have to update prettier to run the app on Node 18 and integrate SDK 1.4.1. It probalby got updated because you removed `yarn.lock` file while installing libs. For such a big transition (from Node 14 to 18) I believe the less changes we made the better. If we want to update some of the libraries, we can always do that in a separate PR.

As of now I recommend to:
1. Revert the `yarn.lock` file to [it's main version](https://github.com/threshold-network/token-dashboard/blob/main/yarn.lock)
2. Revert the changes made in 545a5c32a50eb5a4fbdd123dc9055e35e06d035b
3. In package.json revert `react-icons` to `^4.3.1` and `lint-staged` to `>=10`. As of now those updates are not necessary to migrate the app to Node 18
4. Run `yarn install`
5. Check if the site builds and run properly (I already checked that but it would be good to double check)
6. Push the changes to the branch

kpyszkowski commented:
@michalsmiarowski 
Thanks for analysis. I have:
- reverted 545a5c32a50eb5a4fbdd123dc9055e35e06d035b commit, ref: [7ca2f5d](https://github.com/threshold-network/token-dashboard/pull/652/commits/7ca2f5df41fec260c8f92eb2182b0f0a24b4ea7b)
- checked out `yarn.lock` file to `origin/main` branch, ref: [b8cfd0f](https://github.com/threshold-network/token-dashboard/pull/652/commits/b8cfd0f65a7b9785cd01a0fceaff9538bc37b3fd)
- reverted updated packages, ref: [8972db2](https://github.com/threshold-network/token-dashboard/pull/652/commits/8972db2970849c87c3370107e058c88c8cd3cfef)
- added comment about `@babel/plugin-transform-class-properties` babel plugin, ref: [f5b5358](https://github.com/threshold-network/token-dashboard/pull/652/commits/f5b5358e531c81e2e4c51ebb00f26ef4fac5de5d)

I tested build processes for `goerli` and `development` packages, all works well.

michalsmiarowski commented on craco.config.js:
```diff
@@ -3,6 +3,7 @@ module.exports = {
     plugins: [
       "@babel/plugin-proposal-nullish-coalescing-operator",
       "@babel/plugin-proposal-optional-chaining",
+      ["@babel/plugin-transform-class-properties", { loose: true }],
```
Resolving since it's fixed in f5b5358e531c81e2e4c51ebb00f26ef4fac5de5d

michalsmiarowski reviewed (approved):
LGTM
